### ISSUE TITLE
UseHangfireServer Deprecated - Jobs start before IoC container configured.

### ISSUE THREAD

**Issue by hellblade** [HangfireIO/Hangfire#2117]
With `UseHangfireServer` being obsolete, it can cause an order issue when jobs are running - my IoC container is only populated with the required classes in the `Configure` method.

With `services.AddHangfireServer`, Hangfire starts up before `Configure` is executed so I get failures with dependancies not resolved from IoC.

Is there any way around this, like delaying when Hangfire starts jobs? 
For now I'll have to continue using `UseHangfireServer`.

**Comment by odinserj** [comment]
Could you share your configuration logic? AddHangfireService itself is based on registration logic, so should be called only after everything is configured.

**Comment by hellblade** [comment]
It looks like Configure gets called first but doesn't complete before Hangfire is starting the job.

Note too, it only has an error for jobs that are started right at start up. If a job is triggered after the website is fully started, it runs correctly.

Config is here. I am using ASP.Net Boilerplate if that has anything to do with it.

```
public void ConfigureServices(IServiceCollection services)
{
    services.AddOptions();

    services.AddControllersWithViews(
            options =>
            {
                options.Filters.Add(new AutoValidateAntiforgeryTokenAttribute());
                options.Filters.Add(new AbpAutoValidateAntiforgeryTokenAttribute());
            }
        )
        .AddRazorRuntimeCompilation()
        .AddNewtonsoftJson(options =>
        {
            options.SerializerSettings.ContractResolver = new AbpMvcContractResolver(IocManager.Instance)
            {
                NamingStrategy = new CamelCaseNamingStrategy(),
            };
        });

    IdentityRegistrar.Register(services);
    AuthConfigurer.Configure(services, _appConfiguration);

    services.AddHangfire(config =>
    {
        config.UseSqlServerStorage(_appConfiguration.GetConnectionString("Default"));
    });

    services.AddScoped<IWebResourceManager, WebResourceManager>();

    services.AddSignalR();

    services.AddAbpWithoutCreatingServiceProvider<MyModule>(
        options =>
        {
            options.IocManager.IocContainer.AddFacility<LoggingFacility>(
                f => f.UseAbpLog4Net().WithConfig("log4net.config")
            );

            options.IocManager.IocContainer.Register(Component.For<IConfigurationRoot>().Instance(_appConfiguration));
        }
    );

    services.AddHangfireServer(options =>
    {
        options.WorkerCount = Math.Min(Environment.ProcessorCount * 2, 20);
    });
}

public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILoggerFactory loggerFactory)
{
    app.UseAbp();

    if (env.IsDevelopment())
    {
        app.UseDeveloperExceptionPage();
    }
    else
    {
        app.UseExceptionHandler("/Error");
    }

    app.UseRouting();

    app.UseAuthentication();

    app.UseJwtTokenMiddleware();

    app.UseAuthorization();

    // Works if this is used instead of AddHangfireServer:
    //app.UseHangfireServer(new BackgroundJobServerOptions
    //{
    //    WorkerCount = Math.Min(Environment.ProcessorCount * 2, 20),
    //});

    app.UseHangfireDashboard("/hangfire", new DashboardOptions
    {
        
    });

    app.UseEndpoints(endpoints =>
    {
        // etc
    });
}

**Comment by DamienLaw** [comment]
> Could you share your configuration logic? AddHangfireService itself is based on registration logic, so should be called only after everything is configured.

The thing with ASP.NET Boilerplate is, it is a ASP.NET Core application and the dependency registration is being done in the `Configure()` method of `Startup.cs` instead of in the usual `ConfigureServices()` method. As you already knew, `ConfigureServices()` is run before `Configure()`.

This is fine as long as the Hangfire server is started AFTER the dependency registration. Previously, this can be achieved by placing `app.UseHangfireServer()` after `app.UseAbp();` (this is where all the dependency registrations happen) in the `Configure()` method.

However, `app.UseHangfireServer();` has been deprecated in favour of `services.AddHangfireServer()` which can only be placed in the `ConfigureServices()` method and not in the `Configure()` method. This will cause the Hangfire server to be started before the dependency registration takes place and the cron jobs wouldn't be able to resolve dependencies effectively.

### Solution

Taking inspiration from [Waiting for your ASP.NET Core app to be ready from an IHostedService in .NET 6](https://andrewlock.net/finding-the-urls-of-an-aspnetcore-app-from-a-hosted-service-in-dotnet-6/), we could improve the `BackgroundJobServerHostedService` by injecting a `IHostApplicationLifetime` and modify the method below.
https://github.com/HangfireIO/Hangfire/blob/5b696d4174e13c3dd9489cc6a863d3417c632e31/src/Hangfire.AspNetCore/BackgroundJobServerHostedService.cs#L69-L78

to

```
        public Task StartAsync(CancellationToken cancellationToken)
        {
            _hostApplicationLifetime.ApplicationStarted.Register(() =>
            {
                 _processingServer = _factory != null && _performer != null && _stateChanger != null
#pragma warning disable 618
                    ? new BackgroundJobServer(_options, _storage, _additionalProcesses, null, null, _factory, _performer, _stateChanger)
#pragma warning restore 618
                    : new BackgroundJobServer(_options, _storage, _additionalProcesses);
            });

            return Task.CompletedTask;
        }
```

This way, Hangfire server would only be started when the application is truly ready.

**Comment by DamienLaw** [comment]
> With `UseHangfireServer` being obsolete, it can cause an order issue when jobs are running - my IoC container is only populated with the required classes in the `Configure` method.
> 
> With `services.AddHangfireServer`, Hangfire starts up before `Configure` is executed so I get failures with dependancies not resolved from IoC.
> 
> Is there any way around this, like delaying when Hangfire starts jobs? For now I'll have to continue using `UseHangfireServer`.

I have some luck by placing the Hangfire startup codes into `Program.cs`:
```
public static IHostBuilder CreateHostBuilder(string[] args)
{
    return Host.CreateDefaultBuilder(args)
        .ConfigureWebHostDefaults(webBuilder =>
        {
            webBuilder.ConfigureKestrel(opt =>
            {
                opt.AddServerHeader = false;
                opt.Limits.MaxRequestLineSize = 16 * 1024;
            });
            webBuilder.UseStartup<Startup>();
        })
        .ConfigureServices((context, services) =>
        {
            services.AddHangfire(config =>
            {
                var appConfiguration = context.HostingEnvironment.GetAppConfiguration();
                config.UseSqlServerStorage(appConfiguration.GetConnectionString("Default"));
            });

            services.AddHangfireServer(options =>
            {
                options.WorkerCount = Math.Min(Environment.ProcessorCount * 2, 20);
            });
        });
}
```

More info could be found below

- https://stackoverflow.com/questions/61866319/start-ihostedservice-after-configure#answer-70520190
- https://andrewlock.net/controlling-ihostedservice-execution-order-in-aspnetcore-3/

However, I'd consider this as a hack and it's a coincidence that it works in .NET 6. It might behave differently in future versions.
Without the Hangfire team to implement the `StartAsync()` of the `IHostedService` when the application is ready, you'll have to do it yourself in your cron jobs.

Referring to [https://andrewlock.net/finding-the-urls-of-an-aspnetcore-app-from-a-hosted-service-in-dotnet-6/](https://andrewlock.net/finding-the-urls-of-an-aspnetcore-app-from-a-hosted-service-in-dotnet-6/), you can inject `IHostApplicationLifetime` into your cron job and delay execution of the codes until the application is truly ready.

**Comment by odinserj** [comment]
Thanks a lot for your investigation 👍! I've just added `IHostApplicationLifetime`-based solution, will be released with 1.7.32.
